src/graphql/search/map/normalize.ts

```diff
@@ -0,0 +1,112 @@
+import queryString, { ParsedUrlQueryInput } from 'querystring'
+import { MAX_ZOOM, MIN_ZOOM, MapLayerTypes } from './config'
+import { MapCollection, MapCollectionLayer, LegendItemMapLayer } from '../../../generated/graphql'
+
+type InitialLegendItem = {
+  params: ParsedUrlQueryInput
+} & LegendItemMapLayer
+
+type InitialMapLayer = {
+  minZoom: number
+  maxZoom: number
+  params: ParsedUrlQueryInput
+  legendItems: Array<InitialLegendItem>
+} & MapCollectionLayer
+
+type InitialMapCollection = {
+  mapLayers: Array<InitialMapLayer>
+} & MapCollection
+
+export function composeId(collectionId: string, id: string) {
+  return `${collectionId}-${id}`
+}
+
+export function findMapLayer(mapLayers: Array<InitialMapLayer>, id: string) {
+  return mapLayers.find(mapLayer => mapLayer.id === id)
+}
+
+export function normalizeLegendItems(
+  collectionId: string,
+  legendItems: Array<InitialLegendItem>,
+  mapLayers: Array<InitialMapLayer>,
+) {
+  return legendItems.map(legendItem => {
+    const legendItemLayer = legendItem.id ? findMapLayer(mapLayers, legendItem.id) : null
+    const notSelectable = legendItem.notSelectable || !legendItem.id || false // legendItems with an id are always selectable, unless defined otherwise
+    const params = legendItemLayer?.params || queryString.stringify(legendItem.params) // The field params is an object with unspecified content, it's stringefied here to make typing easier
+
+    return {
+      ...(legendItemLayer || legendItem), // If a matching mapLayer is found, this data should be used
+      ...(legendItem.type ? { type: MapLayerTypes[legendItem.type] } : {}), // Get the legendItem type from constants
+      imageRule: legendItem.imageRule || legendItemLayer?.imageRule, // imageRule can be overwritter per collection and mapLayer
+      title: legendItem.title || legendItemLayer?.title, // title can be overwritter per collection and mapLayer
+      // The ID of the mapLayer when defined as legendItem, is a combination of the IDs of the mapLayer and the collection it's used in to prevent duplication
+      id: !notSelectable && legendItem.id ? composeId(collectionId, legendItem.id) : null, // Only selectable legendItems need an id
+      notSelectable,
+      noDetail: !legendItem.detailUrl,
+      params,
+    }
+  }) as Array<LegendItemMapLayer>
```

Comment:
By casting the value here we lose a lot of type safety checks of the object as we're basically telling TypeScript 'I am sure this object is of this type' eliminating some checks the compiler can do for us.

I'd recommend adding this type to the return value of the method instead and fixing any errors that might show up.